Multi-hit refactor and Parental Bond

BuffelSaft opened the pull request:
Port multi-hit system from the CFRU, implement Parental Bond, refactor certain moves as required and implement all remaining multi-hit moves except Dragon Darts and Scale Shot.

## Description
This PR does the following:
- Implements Parental Bond
- Replaces the existing script-based multi hit system with the CFRU's version, which sets up the number of hits in the attack canceller and executes the follow up hits as a looping move end effect.
- Implements the following move effects: Triple Axel, Surging Strikes, Double Iron Bash
- Implements Triple Kick/Skill Link interaction (no accuracy check on follow up hits if if user has Skill Link)
- Updates Beat Up to its modern effect and adds a config option to switch between this and its old effect
- Refactors the following move effects to be move end effects (like Knock Off) so that they happen after both strikes of Parental Bond: 
  - Smack Down
  - Burn Up
  - Wake Up Slap/Smelling Salts/Sparkling Aria
- Adds FLAG_TWO_STRIKES. Any move with this flag will strike twice and its secondary effect will occur on both strikes (see the changes to Twineedle for an example).

The following Parental Bond-affected edge cases are accounted for:
- Fury Cutter and Echoed Voice increase their damage counter on the second strike of Parental Bond, so that they don't get boosted damage after the first hit.
- If Present heals the foe, it doesn't strike twice. If Present damages the foe, the second hit will always do damage.
- Knock Off and Wake-Up Slap deal bonus damage on both strikes, as their effects happen after both hits
- Burn Up receives STAB on both hits
- Smack Down knocks down the target after both hits
- Items are stolen after both hits, so items can activate after the first hit (and not be stolen) if applicable
- Recoil damage happens after both strikes and is based on the total damage of both hits
- Bug Bite's berry stealing effect happens after the final hit

Some things aren't handled yet:
- Relic Song should change Meloetta's form after both hits if it has Parental Bond (Relic Song effect not done yet)
- Secret Power's secondary effect happens only on the second strike. I had some trouble with this one, and I'm not sure if it can just be made into a move end effect like the Smack Down et al.
- Struggle does not deal 50% of the user's HP if it has Parental Bond. This is because Struggle has not been updated to its modern effect, so that can be done in a separate PR.

Notes for testing:
I've had to rearrange the order that some move end effects happen, and while it seems fine in my testing so far, it's a major change and will need thorough testing.

Gifs:
Recharge moves and fixed damage moves:
![PB recharge fixed damage](https://user-images.githubusercontent.com/28769716/134447593-21c989e9-e6b2-4fd9-b6d2-9700991cfbfc.gif)
Recoil after both hits (Double-Edge changed to fighting type for testing):
![PB Recoil](https://user-images.githubusercontent.com/28769716/134447600-0e52457b-b248-45cc-8235-b2900a63a0ab.gif)
Target smacked down after both hits:
![PB Smack Down](https://user-images.githubusercontent.com/28769716/134447604-9e406a4e-7d7c-4285-ae43-498936dfa036.gif)
Spread moves hit once if there are two targets, twice otherwise:
![PB spread move 2 targets](https://user-images.githubusercontent.com/28769716/134447606-96d3529a-a830-4cc8-87fb-da3fa74e0219.gif)
Status is cured after both hits for Smelling Salts etc.:
![PB cure status](https://user-images.githubusercontent.com/28769716/134447590-5c438ce1-5bbb-493e-a3c8-58124d214568.gif)
Knock Off removes items after both hits:
![PB knock off](https://user-images.githubusercontent.com/28769716/134447592-dc69c3fe-1e39-480c-8815-9159c752a6b9.gif)
Burn Up removes Fire typing after both hits:
![PB burn up](https://user-images.githubusercontent.com/28769716/134447583-f93c81ac-a97b-48ec-bd27-364e787a22dc.gif)

Addresses the following issues:
#351
#1102 
#1107 
#1402 

## **Discord contact info**
Buffel Saft#2205

LOuroboros commented:
There's conflicts that need to be solved :eyes:

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -1669,6 +1717,13 @@ static void Cmd_accuracycheck(void)
             gBattlescriptCurrInstr += 7;
     }
+    else if (gSpecialStatuses[gBattlerAttacker].parentalBondOn == 1
+		|| (gSpecialStatuses[gBattlerAttacker].multiHitOn && (gBattleMoves[move].effect != EFFECT_TRIPLE_KICK
+		|| GetBattlerAbility(gBattlerAttacker) == ABILITY_SKILL_LINK)))
```
Spaces instead of tabs.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -2008,6 +2064,12 @@ static void Cmd_attackanimation(void)
     {
+        if (gSpecialStatuses[gBattlerAttacker].parentalBondOn == 1) // No animation on second hit
+        {
+			gBattlescriptCurrInstr++;
+			return;
```
Spaces instead of tabs.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -5076,6 +5125,56 @@ static void Cmd_moveend(void)
+                gBattleScripting.multihitString[4]++;
+                if (--gMultiHitCounter == 0)
+                {
+					BattleScriptPushCursor();
+					gBattlescriptCurrInstr = BattleScript_MultiHitPrintStrings;
+					effect = 1;
```
Spaces instead of tabs.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -5076,6 +5125,56 @@ static void Cmd_moveend(void)
+                    if (gBattleMons[gBattlerAttacker].hp
+                    && gBattleMons[gBattlerTarget].hp
+                    && (gChosenMove == MOVE_SLEEP_TALK || !(gBattleMons[gBattlerAttacker].status1 & STATUS1_SLEEP))
+					&& !(gBattleMons[gBattlerAttacker].status1 & STATUS1_FREEZE))
```
Spaces instead of tabs.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -5076,6 +5125,56 @@ static void Cmd_moveend(void)
+                        gSpecialStatuses[gBattlerAttacker].multiHitOn = TRUE;
+                        MoveValuesCleanUp();
+                        BattleScriptPush(gBattleScriptsForMoveEffects[gBattleMoves[gCurrentMove].effect]);
+						gBattlescriptCurrInstr = BattleScript_FlushMessageBox;
+						return;
```
Spaces instead of tabs.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -5076,6 +5125,56 @@ static void Cmd_moveend(void)
+                    else
+                    {
+                        BattleScriptPushCursor();
+						gBattlescriptCurrInstr = BattleScript_MultiHitPrintStrings;
+						effect = 1;
```
Spaces instead of tabs.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -12873,3 +12983,27 @@ static bool32 CriticalCapture(u32 odds)
+bool8 IsMoveAffectedByParentalBond(u16 move, u8 battlerId)
+{
+    if (gBattleMoves[move].split != SPLIT_STATUS && !(sForbiddenMoves[move] & FORBIDDEN_PARENTAL_BOND))
+	{
+		if (gBattleTypeFlags & BATTLE_TYPE_DOUBLE)
+		{
+			switch (gBattleMoves[move].target)
```
Spaces instead of tabs.

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -12873,3 +12983,27 @@ static bool32 CriticalCapture(u32 odds)
+		{
+			switch (gBattleMoves[move].target)
+            {
+				case MOVE_TARGET_BOTH:
+					if (CountAliveMonsInBattle(BATTLE_ALIVE_DEF_SIDE) >= 2) // Check for single target
+						return FALSE;
+					break;
+				case MOVE_TARGET_FOES_AND_ALLY:
+					if (CountAliveMonsInBattle(BATTLE_ALIVE_EXCEPT_ACTIVE) >= 2) // Count mons on both sides; ignore attacker
+						return FALSE;
+					break;
+			}
+		}
+
+		return TRUE;
+	}
+
+	return FALSE;
```
Spaces instead of tabs.

LOuroboros commented on src/battle_util.c:
```diff
@@ -3412,6 +3446,95 @@ u8 AtkCanceller_UnableToUseMove(void)
+            else if (IsTwoStrikesMove(gCurrentMove))
+            {
+                gMultiHitCounter = 2;
+				PREPARE_BYTE_NUMBER_BUFFER(gBattleScripting.multihitString, 1, 0)
```
Spaces instead of tabs. Also please align the value setting of `gMultiHitCounter` with the call to the `PREPARE_BYTE_NUMBER_BUFFER` macro.

LOuroboros commented on src/battle_util.c:
```diff
@@ -3412,6 +3446,95 @@ u8 AtkCanceller_UnableToUseMove(void)
+                    party = gEnemyParty;
+                
+                for (i = 0; i < PARTY_SIZE; i++)
+				{
+					if (GetMonData(&party[i], MON_DATA_HP)
+					&& GetMonData(&party[i], MON_DATA_SPECIES) != SPECIES_NONE
+					&& !GetMonData(&party[i], MON_DATA_IS_EGG)
+					&& !GetMonData(&party[i], MON_DATA_STATUS))
+						gMultiHitCounter++;
+				}
+
+				gBattleCommunication[0] = 0; // For later
+				PREPARE_BYTE_NUMBER_BUFFER(gBattleScripting.multihitString, 1, 0)
```
Spaces instead of tabs.

LOuroboros commented on src/battle_util.c:
```diff
@@ -8163,6 +8291,15 @@ static u32 CalcFinalDmg(u32 dmg, u16 move, u8 battlerAtk, u8 battlerDef, u8 move
     }
 
+    // Parental Bond Second Strike
+	if (gSpecialStatuses[gBattlerAttacker].parentalBondOn == 1)
+	{
+		if (B_PARENTAL_BOND_DAMAGE < GEN_7)
+			MulModifier(&finalModifier, UQ_4_12(0.5));
+		else
+			MulModifier(&finalModifier, UQ_4_12(0.25));
+	}
```
Spaces instead of tabs.

BuffelSaft commented:
Conflicts resolved and tabs replaced!

LOuroboros commented:
More conflicts to solve :eyes:

LOuroboros commented on src/battle_util.c:
```diff
@@ -7891,11 +7880,14 @@ static u16 CalcMoveBasePower(u16 move, u8 battlerAtk, u8 battlerDef)
         #if B_BEAT_UP_DMG >= GEN_5
         basePower = CalcBeatUpPower();
         #endif
-        break;
```
This `break;` should probably be added back.

BuffelSaft replied:
Done!

AsparagusEduardo reviewed (changes requested):
Here are my notes:
- You need to merge with BE in order to check if Relic Song is working with Parental Bond.
- There's an infinite loop when using bide with Parental Bond
![bideloop](https://user-images.githubusercontent.com/2904965/140607957-ac922fa9-d9fd-4f8c-9c83-0003f0293123.gif)
- Earthquake and Surf should hit twice when it can only hit one target
![earthquake](https://user-images.githubusercontent.com/2904965/140608057-65cbebac-84ea-4ab2-a0d6-1c9a6bd5a463.gif) ![surf](https://user-images.githubusercontent.com/2904965/140608061-1e7d2b37-5442-4c44-b040-7c111d3c06b4.gif) ![earthsingle](https://user-images.githubusercontent.com/2904965/140608094-03eb7b30-5110-4562-b28d-e06e9a0630b6.gif)
- Secret Power's secondary effect should only trigger after the 2nd strike (right now it's doing it for both). Seeing as Smack Down, Smelling Salts, Knock Off and Burn Up are all handled in the same place and their effect is being triggered after the 2nd hit, it might be good to see if we can do the same for Secret Power.
![secretpower](https://user-images.githubusercontent.com/2904965/140608737-f4452d6e-cc24-4ffe-8308-46d71f2fe352.gif)
- Payday should only scatter coins on the first turn. (Without ParentalBond, it's getting 30$, so its definetly doing the effect twice).
![payday](https://user-images.githubusercontent.com/2904965/140608881-2f55c741-2065-46db-826c-d44259e26098.gif)
- Struggle is not dealing 50% of the user's HP now that #1710 has been merged
![struggle](https://user-images.githubusercontent.com/2904965/140609262-b0a900cb-85a9-4447-83ac-5f486d687a0c.gif).
- Dragon Tail and Circle Throw should switch the target after the 2nd hit.
![dragontail](https://user-images.githubusercontent.com/2904965/140609798-5a5757ad-0d74-4437-8f7f-264ce9d6fed7.gif)
- Natural Gift and Spit Up should hit twice.
![naturalgift](https://user-images.githubusercontent.com/2904965/140610392-905d55fc-dbc8-42ad-a124-c2c9aa70af76.gif) ![spitup](https://user-images.githubusercontent.com/2904965/140610452-0d5bb000-0b8e-4867-9c33-59b9be5233a1.gif)

BuffelSaft commented:
Earthquake is working on my local, but the other issues are definitely not handled currently. I need to rework the way I've done the once-per-turn effects, it's causing some issues with Burn Up, and it might cause headaches with future move effects.

BuffelSaft commented:
I think I've fixed all the reported issues, but I've having trouble getting Bide to strike twice on its damaging turn (the way I fixed the loop means it's basically unaffected by Parental Bond, need to make some changes to Bide itself). Other than that this is ready for re-review.

kleeenexfeu commented on src/battle_script_commands.c:
```diff
@@ -13929,3 +14112,43 @@ static bool32 CriticalCapture(u32 odds)
 }
+
+bool8 IsMoveAffectedByParentalBond(u16 move, u8 battlerId)
+{
```
Takes u8 battlerId in parameter but never uses it; I think it can be removed

LOuroboros commented on src/battle_script_commands.c:
```diff
@@ -13929,3 +14112,43 @@ static bool32 CriticalCapture(u32 odds)
+        {
+            switch (gBattleMoves[move].target)
+            {
+                case MOVE_TARGET_BOTH:
+                    if (CountAliveMonsInBattle(BATTLE_ALIVE_DEF_SIDE) >= 2) // Check for single target
+                        return FALSE;
+                    break;
+                case MOVE_TARGET_FOES_AND_ALLY:
+                    if (CountAliveMonsInBattle(BATTLE_ALIVE_EXCEPT_ACTIVE) >= 2) // Count mons on both sides; ignore attacker
+                        return FALSE;
+                    break;
```
```suggestion
            case MOVE_TARGET_BOTH:
                if (CountAliveMonsInBattle(BATTLE_ALIVE_DEF_SIDE) >= 2) // Check for single target
                    return FALSE;
                break;
            case MOVE_TARGET_FOES_AND_ALLY:
                if (CountAliveMonsInBattle(BATTLE_ALIVE_EXCEPT_ACTIVE) >= 2) // Count mons on both sides; ignore attacker
                    return FALSE;
                break;
```

AsparagusEduardo commented:
Re tested after `make clean` and Earthquake is still not being affected when there's only 1 target.
![2WxxV1n7Sw](https://user-images.githubusercontent.com/2904965/146652862-38435a23-7615-4b45-9499-52ee289940af.gif)

BuffelSaft commented:
Ready for re-re-...re-review! Note that Bide does not work correctly with Parental Bond. I'll handle that separately as it looks like Bide will need some significant changes to make it work. It's currently prevented from striking twice to prevent the infinite loop that @AsparagusEduardo found.

BuffelSaft commented:
![Earthquake Wingull](https://user-images.githubusercontent.com/28769716/197322924-234f419d-cedd-43a3-a299-6094400ec945.gif)
![Earthquake](https://user-images.githubusercontent.com/28769716/197322928-4c857c37-9f51-4569-9ea0-5fb37a21c0c4.gif)

@AsparagusEduardo Ready for review! I made sure to `make clean` and check that Earthquake is working. Hopefully it works on your end too.

AsparagusEduardo reviewed (changes requested):
### Parental Bond notes

Cases:
- [x] Causes most damaging moves used by the Pokémon to become two-strike moves with the second strike having its damage halved
    - [x] Unless it is a set-damage move (such as Seismic Toss and Dragon Rage), which deal the full amount of damage for both strikes
    - [x] Check Gen 6 damage multiplier.
- [x] It does not affect multi-strike moves
- [x] If a move could hit multiple targets (including allies), such as Earthquake and Rock Slide, it will not strike twice
    - [x] If a move could hit multiple Pokémon but only hits one due to missing the other Pokémon, it will only strike once.
    - [x] however, if it can only hit a single Pokémon, such as in a Single Battle it will strike twice.
- [x] Since Parental Bond turns moves into two-hit multi-strike moves
    - [x] each strike has a separate chance to be a critical hit
    - [x] items that trigger upon strike or contact such as Rocky Helmet occur for each strike
    - [x] Abilities that trigger upon strike or contact such as Cursed Body occur for each strike
- [x] there is only one accuracy check, so either both strikes hit or both strikes miss.
- [x] Any attack which has a secondary effect has the same secondary effect on both strikes (such as Power-Up Punch);
    - [x] except Secret Power's secondary effect that can only occur after the final strike.
    - [x] each strike has an independent chance of activating that effect.
- [x] Even if the Pokémon's Ability is changed to Mummy after the first strike, it will continue to make a second strike regardless.
- [x] Pay Day scatters coins after the first strike only.
- [x] Incinerate destroys applicable held items after each strike.
- [x] If a move has recoil damage, the recoil will be based on the damage dealt by both strikes, but will be taken after the final strike.
- [x] Struggle will inflict recoil damage equal to half the user's maximum HP (after the final strike)
- [x] Moves that switch the user out strike twice, then force a Pokémon to switch out after both strikes are conducted.
- [x] Thief, Covet, Bug Bite, and Pluck do not steal or eat the target's held item until after the final strike, so if the target could use its item after the first strike (e.g. due to low HP), it will use it before the attacker can steal or eat it.
- [x] Smelling Salts, Wake-Up Slap, and Knock Off do not cure the target's status condition or remove its held item (respectively) until after the final strike, so both strikes get the increased power.
- [x] Fire-type moves, Scald, and Steam Eruption thaw a frozen target after the final strike (so a frozen target cannot be thawed and then burned by the same move).
- [x] Smack Down and Thousand Arrows only cause the target to fall to the ground after the final strike.
- [x] If Meloetta has Parental Bond and uses Relic Song, it will change Forme only once, after the final strike.
- [x] Burn Up does not remove the user's Fire type until after the second strike (so both strikes receive same-type attack bonus).
- [x] If Present heals the target it will only strike once, but if it damages the target it will strike twice (the second strike will always damage the target).
- [x] The damage dealt by Psywave is generated separately for each strike, and the second strike's damage is not halved.
- [x] Each strike of Super Fang halves the target's HP (effectively quartering it if HP is not changed between strikes).
- [x] Counter, Mirror Coat, Metal Burst deal the full amount of damage for both strikes.
- [x] The first strike of Assurance counts as previously taking damage for the second strike, giving it increased power.
- [x] Fury Cutter and Echoed Voice only consider uses of the move rather than hits, so the second strike's power is not boosted by the first strike.
- [x] Natural Gift strikes twice.
- [x] Spit Up strikes twice.
- [x] Moves that require recharging after use strike twice, but the user only needs to recharge for one turn.
- [x] One-hit knockout moves only strike once.
- [x] Fling only strikes once.
- [x] Self-Destruct and Explosion only strike once.
- [x] Final Gambit only strikes once.
- [x] Uproar only strikes once.
- [x] Rollout and Ice Ball only strike once.
    - [x] Other consecutively executed moves, such as Outrage, can strike twice.
- [x] Moves with a charging turn (such as Fly and Solar Beam) only strike once, even if the Pokémon becomes fully charged in one turn (such as with a Power Herb).
- [x] Endeavor also only strikes once, even if the user or target's HP is changed after it strikes (such as by Iron Barbs or the Sitrus Berry).
- [x] Confusion damage only occurs once.
- [x] Spirit Shackle and Anchor Shot only trap the target after the final strike.

Failed tests
- [ ] If using a move that can hit multiple targets when only one target is in range, it will strike twice.
![earthquake](https://user-images.githubusercontent.com/2904965/197350482-ac49e5dd-3ab7-42c5-920a-4e69450b284c.gif)

- [ ] Spiky Shield and King's Shield only damage and decrease Attack (respectively) once if they protect a Pokémon from a contact move used by a Pokémon with Parental Bond.
![kingshield](https://user-images.githubusercontent.com/2904965/197350602-6a0cd5c9-b99e-42e6-bacd-deb9b78cedc3.gif) ![spikyshield](https://user-images.githubusercontent.com/2904965/197350604-19cff24e-a258-4899-aa7f-3c51848ad4e4.gif)

- [ ] Moves that switch the target out strike twice, then force a Pokémon to switch out after both strikes are conducted.
![dragontail](https://user-images.githubusercontent.com/2904965/197350641-e454b4d2-ca51-4fb1-85e4-6629cb18e4fb.gif)

- [ ] Parental Bond does not affect Z-Moves.
![zmove](https://user-images.githubusercontent.com/2904965/197350662-a98a47a4-9944-402c-ad78-6a251e5ee86a.gif)

Suggested improvements:
- [ ] Using constants for parentalBondOn (PARENTAL_BOND_1ST_HIT, PARENTAL_BOND_2ND_HIT, PARENTAL_BOND_OFF). Rename parentalBondOn to `parentalBondState`.
- [ ] Adding a function to get the random 2-5 multi-hit count.
    - [ ] use `ifdef` instead of `if`
- [ ] Self-destruct and Explosion don't need the forbidden flag, as the user faints before a second hit, likely canceling the second strike. Should we fix this so that they do hit twice and they need the flag? :eyes:
    - And if we do, we'll need to add Misty Explosion to the ban list.

Not tested, requires unimplemented feature:
- [ ] Grass Pledge, Fire Pledge, and Water Pledge strike twice, even when used as a combination move.
- [ ] Parental Bond does not affect Max Moves.

To be fixed in a later PR:
- [ ] Bide deals the full amount of damage for both strikes.

### Beat Up Notes:
- [ ] There's a noticable delay between hits, in both pre and post Gen5 configs.
https://user-images.githubusercontent.com/2904965/197350698-434ff4c6-a1c7-4f5d-9986-541e677b49ca.mp4


Suggestion:
- [ ] B_UPDATED_MOVE_DATA for Beat Up should be replaced by B_BEAT_UP_DMG.

BuffelSaft commented:
@AsparagusEduardo I've fixed the following:
- If using a move that can hit multiple targets when only one target is in range, it will strike twice (but actually fixed this time, I hope!)
- Spiky Shield and King's Shield only damage and decrease Attack (respectively) once if they protect a Pokémon from a contact move used by a Pokémon with Parental Bond.
- Moves that switch the target out strike twice, then force a Pokémon to switch out after both strikes are conducted.
- Parental Bond does not affect Z-Moves.
I've also changed parentalBondOn to parentalBondState, added constants for each state, and moved the random hit counter to a separate function.

I left Explosion in the ban list just in case something wacky gets done to it in future; this way it will always work properly with Parental Bond. IMO it's also nice to have all the unaffected moves documented in the code.

Oddly I'm not seeing the Beat Up delay in the gen 5+ version, though it should be there for gen 3 and 4 as the vanilla version waits ages for the "mon's attack!" string on each hit.

I think Beat Up is worth a separate config because it's essentially a whole new move effect with very different mechanics, rather than a simple data change like most updated moves get. Could move it to other move settings if that fits better?

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -14844,12 +14844,16 @@ bool8 IsMoveAffectedByParentalBond(u16 move, u8 battlerId)
+            // Both foes are alive, spread move strikes once
             case MOVE_TARGET_BOTH:
-                if (CountAliveMonsInBattle(BATTLE_ALIVE_DEF_SIDE) >= 2) // Check for single target
+                if (IsBattlerAlive(gBattlerTarget) && IsBattlerAlive(BATTLE_PARTNER(gBattlerTarget)))
```
I did a test with the original code and it worked for `MOVE_TARGET_BOTH`. I think I wasn't clear enough to say that the issue I was having was with `MOVE_TARGET_FOES_AND_ALLY` moves.

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -14844,12 +14844,16 @@ bool8 IsMoveAffectedByParentalBond(u16 move, u8 battlerId)
+            // Either both foes or one foe and its ally are alive; spread move strikes once
             case MOVE_TARGET_FOES_AND_ALLY:
-                if (CountAliveMonsInBattle(BATTLE_ALIVE_EXCEPT_ACTIVE) >= 2) // Count mons on both sides; ignore attacker
+                if (IsBattlerAlive(BATTLE_PARTNER(gBattlerTarget))
```
I made a new condition for `CountAliveMonsInBattle` which seems to work:
```
    case BATTLE_ALIVE_EXCEPT_ATTACKER:
        for (i = 0; i < MAX_BATTLERS_COUNT; i++)
        {
            if (i != gBattlerAttacker && !(gAbsentBattlerFlags & gBitTable[i]))
                retVal++;
        }
        break;
```
![exceptAttacker](https://user-images.githubusercontent.com/2904965/197428156-67108128-8767-4b2b-9030-24198b1b252e.gif) ![exceptAttacker2](https://user-images.githubusercontent.com/2904965/197428158-46f0c372-dd61-4515-9fb4-00d9f1376f1b.gif)

AsparagusEduardo reviewed (changes requested):
The pause seems to be happening for every multi-hit iteration, not only Beat Up. It's between the times the mon takes damage.
Are you sure you don't notice it?

https://user-images.githubusercontent.com/2904965/197427957-6a325d22-b5c2-46fb-8bdd-86a7a96c33e2.mp4

BuffelSaft commented:
> The pause seems to be happening for every multi-hit iteration, not only Beat Up. It's between the times the mon takes damage. Are you sure you don't notice it?
> 
>  PauseMultiHit.mp4

I can definitely see it on Parental Bond, but not on other multihit moves:
![PB pause](https://user-images.githubusercontent.com/28769716/197432009-a4fa5ce5-b6cd-4adb-96ae-19a2f5c3a906.gif)
![pause](https://user-images.githubusercontent.com/28769716/197432013-fefaea8b-4b63-4d77-a26f-ae71a5bd68fb.gif)

BuffelSaft commented:
Suggested changes done!

AsparagusEduardo commented:
> I can definitely see it on Parental Bond, but not on other multihit moves

Yeah, that was my bad wording. I meant to say PB moves were also affected.

AsparagusEduardo commented:
> I think Beat Up is worth a separate config because it's essentially a whole new move effect with very different mechanics, rather than a simple data change like most updated moves get. Could move it to other move settings if that fits better?

I forgot to say that I agree with this.

BuffelSaft commented:
I've moved `B_BEAT_UP_DMG` to other move settings. Can't see anything that would be causing the pause between hits, as that should just be flushing the message box like multi target moves do. Could tackle that as a separate issue if you don't mind? It's a little annoying but not game breaking.

AsparagusEduardo commented on include/battle.h:
```diff
@@ -163,15 +163,20 @@ struct SpecialStatus
     u8 stormDrainRedirected:1;
     u8 switchInAbilityDone:1;
-    u8 switchInItemDone:1;
+    u8 switchInItemDone:1; 
```
Trailing whitespace.

AsparagusEduardo commented on include/constants/battle_config.h:
```diff
@@ -89,6 +90,7 @@
 #define B_ROOTED_GROUNDING          GEN_LATEST // In Gen4+, Ingrain causes the affected Pokémon to become grounded.
 #define B_METRONOME_MOVES           GEN_LATEST // This config will determine up to which generation will Metronome pull moves from.
 #define B_TELEPORT_BEHAVIOR         GEN_LATEST // In Gen7+, starting with Pokémon LGPE, Teleport allows the user to swap out with another party member.
+#define B_BEAT_UP_DMG               GEN_LATEST // In Gen5+, Beat Up uses a different formula to calculate the damage of each hit, and deals Dark-type damage.
```
I'm thinking we should rename the define to reflect that it's not only the damage. `B_BEAT_UP` should be enough. Also, better add in that it announces mon names pre-Gen5

AsparagusEduardo commented on include/constants/battle_config.h:
```diff
@@ -29,6 +29,7 @@
 #define B_KNOCK_OFF_DMG             GEN_LATEST // In Gen6+, Knock Off deals 50% more damage when knocking off an item.
 #define B_SPORT_DMG_REDUCTION       GEN_LATEST // In Gen5+, Water/Mud Sport reduce Fire/Electric Damage by 67% instead of 50%.
 #define B_EXPLOSION_DEFENSE         GEN_LATEST // In Gen5+, Self-Destruct and Explosion don't halve the targets' defense.
+#define B_PARENTAL_BOND_DAMAGE      GEN_LATEST // In Gen7+, Parental Bond's second hit does 25% of the initial hits damage. Before, it did 50%.
```
To keep consistent with the other settings, `_DMG` would be better.

AsparagusEduardo commented:
I will take a look at look at the delay issue later today. After that and the other changes I requested now, we should be good to go :D

BuffelSaft commented:
Changes done!

AsparagusEduardo commented:
So far I've noticed that the delay is happening on double battles when there's only the mon with PB and the target.

AsparagusEduardo commented:
Found a fix to the delay. It seems to be happening every time `stringId` not set, not only for multi-hit moves with no effectiveness message:

```diff
    if (stringId)
+   {
        PrepareStringBattle(stringId, gBattlerAttacker);
+       gBattlescriptCurrInstr++;
+   }
+   else
+       gBattlescriptCurrInstr += 2; //Skips the following waitmessage, as there's nothing to print.
-   gBattlescriptCurrInstr++;
```

AsparagusEduardo reviewed (approved):
Seeing as the only missing thing is that delay fix that we already found a solution for,
I'll call this PR complete and submit the fix separetly, as it'll be better than risking getting new conflicts.
